# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

# 读取数据

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv("e_commerce.csv")
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

# 评估数据

## 评估数据整洁度

In [3]:
original_data.sample(10)

InvoiceNo StockCode                         Description  Quantity  \
463483    576064     22619           SET OF 6 SOLDIER SKITTLES         2   
52501     540798     85197  SET OF 12 MINI BUNNIES IN A BUCKET         3   
330670    565920     22066              LOVE HEART TRINKET POT        48   
27545     538566     22178     VICTORIAN GLASS HANGING T-LIGHT         2   
291104    562427     84817        DANISH ROSE DECORATIVE PLATE         2   
421376    572960     22601      CHRISTMAS RETROSPOT ANGEL WOOD        24   
5014      536836     22649        STRAWBERRY FAIRY CAKE TEAPOT         1   
87868     543734    84520B  PACK 20 ENGLISH ROSE PAPER NAPKINS        12   
227163    556824     22839     3 TIER CAKE TIN GREEN AND CREAM         1   
46986     540372     20679               EDWARDIAN PARASOL RED         3   

             InvoiceDate  UnitPrice  CustomerID         Country  
463483  11/13/2011 15:37       3.75     12508.0          France  
52501    1/11/2011 12:11       2.10     17365.0  United Kingdom  
330670     9/8/2011 8:56       0.39     15298.0  United Kingdom  
27545   12/13/2010 11:21       2.51         NaN  United Kingdom  
291104    8/4/2011 17:19       2.10     17267.0  United Kingdom  
421376  10/27/2011 10:40       0.85     18154.0  United Kingdom  
5014     12/2/2010 18:08       4.95     18168.0  United Kingdom  
87868    2/11/2011 12:22       0.85     17512.0  United Kingdom  
227163   6/15/2011 10:24      29.17         NaN  United Kingdom  
46986     1/6/2011 16:41       5.95     13081.0  United Kingdom

##### 从抽样的10条数据来看，数据结构符合每列一个变量，每行一个观察值，每个单元格是一个值。具体来说每行是一个商品的一次交易记录，每列是交易相关的各个变量，不存在数据结构问题

## 评估数据干净度

In [4]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


##### 数据一共有541909条，Description，CustomerID存在缺失值，InvoiceDate的数据类型应为日期，CustomerID的数据类型硬是字符串，应该进行数据格式转换。

## 评估缺失数据

### 根据条件抽取出"Description"有缺失值的观察值

In [5]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

#### 从运行结果来看，Description为空的数据有1454条 
观察这些数据，发现Description为空的同时UnitPrice=0，为了验证这个猜想，增加筛选条件，看是否存在Description为空并且nitPrice不是0的数据

In [6]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

##### 筛选结果为0，说明缺失Description值的数据，同时也不具备有效的UnitPrice 
##### 我们要分析的是畅销商品的数据，商品名称（Description）和商品价格（UnitPrice）都是进行后续分析的重要变量，如果值缺失则不能提供有效含义，后续可以进行删除。

In [7]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

##### CustomerID是顾客编号，并不是分析畅销商品的必要变量，从CustomerID有缺失值的观察数据来看，有些交易仍然有效，因此保留此缺失值的数据

## 评估重复数据

##### InvoiceNo,StockCode,CustomerID虽然是唯一标识符，但一次交易可以有多个商品，所以InvoiceNo可以重复，不同交易可以有同一个商品，同一个顾客也可以购买不同商品，所以对于数据集无需评估重复数据

## 评估不一致数据

### 不一致数据可能存在于Country中，看看是否存在不同内容表示同一国家的情况

In [8]:
original_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


##### 从统计结果来看，U.K. ，UK，United Kingdom都表示英国，United States，USA都是美国，清洗数据时需要对这些值进行统一

## 评估无效或错误数据

### 通过DataFrame的describe方法，对数值统计信息快速了解

In [9]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

#### 筛选出Quantity是负数的观察值

In [10]:
original_data[original_data["Quantity"] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

##### 判断Quantity为负数的观察值，是否InvoiceNo都是以"C"开头的

In [11]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

##### Quantity是负数，并且InvoiceNo不是"C"开头的观察值中，是否存在UnitPrice不为0的观察值

In [12]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C") & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

##### 经过上记筛选后，发现Quantity为负数的观察值满足下记两种情况 
- 发票编号(InvoiceNo)以字母"C"开头，表示订单被取消
- 商品单价(UnitPrice)是0
##### 这些交易数据不是正规的销售数据，所以不用做分析，可以把Quantity为负数的观察值删掉

#### 筛选出UnitPrice是负数的观察值

In [13]:
original_data[original_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

##### 从输出来看，Adjust bad debt是坏账调整，也不是正常的交易数据，可以删除掉

# 清洗数据

##### 根据前面评估部分得到的结论，我们需要进行的数据清理包括： 

- 把InvoiceDate变量的数据类型转换为为日期
- 把CustomerID变量的数据类型转换为字符串
-把Description变量缺失的观察值删除
- 把Country变量值"USA"替换为"United States"
- 把Country变量值"UK"、"U.K."替换为"United Kingdom"
- 把Quantity变量值为负数的观察值删除
- 把UnitPrice变量值为负数的观察值删除负数的观察值删除负数的观察值删除

##### 为了区分开经过清理的数据和原始的数据，我们创建新的变量cleaned_data，让它为original_data复制出的副本。我们之后的清理步骤都将被运用在cleaned_data上。

In [14]:
cleaned_data = original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [15]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])

In [16]:
cleaned_data.sample(3)

InvoiceNo StockCode                    Description  Quantity  \
403237    571556     20726             LUNCH BAG WOODLAND        10   
373740    569360     23165  LARGE CERAMIC TOP STORAGE JAR         8   
51362     540646     21012  ANTIQUE ALL GLASS CANDLESTICK         5   

               InvoiceDate  UnitPrice  CustomerID         Country  
403237 2011-10-18 09:24:00       1.65     12662.0         Germany  
373740 2011-10-03 14:59:00       1.65     14653.0  United Kingdom  
51362  2011-01-10 14:32:00       2.51         NaN  United Kingdom

In [17]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)

In [18]:
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

##### 把CustomerID变量值结尾的".0"删除：

In [19]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0,-2)

In [20]:
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

In [23]:
cleaned_data.sample(2)

InvoiceNo StockCode                      Description  Quantity  \
464858    576189     23493  VINTAGE DOILY TRAVEL SEWING KIT         2   
148937    549261     22727        ALARM CLOCK BAKELIKE RED          3   

               InvoiceDate  UnitPrice CustomerID         Country  
464858 2011-11-14 11:33:00       1.95      14684  United Kingdom  
148937 2011-04-07 12:35:00       3.75      14903  United Kingdom

In [25]:
cleaned_data.dropna(subset=["Description"],inplace=True)

In [26]:
cleaned_data["Description"].isnull().sum()

0

In [27]:
cleaned_data["Country"] = cleaned_data["Country"].replace("USA","United States")

In [28]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

In [29]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK":"United Kingdom","U.K.":"United Kingdom"})

In [31]:
len(cleaned_data[(cleaned_data["Country"] == "UK") | (cleaned_data["Country"] == "U.K.")])

0

In [32]:
cleaned_data["Country"].value_counts()

Country
United Kingdom          494024
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
United States              291
China                      288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


In [34]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]

In [35]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

In [36]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]

In [37]:
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

# 保存清洗后的数据

### 完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为e_commerce_cleaned.csv。

In [38]:
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom

In [39]:
cleaned_data.to_csv("e_commerce_cleaned.csv",index=False)

In [40]:
pd.read_csv("e_commerce_cleaned.csv").head()

C:\Users\221081672\AppData\Local\Temp\ipykernel_5168\2686214642.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom